### PR TITLE
feat: new entity in storage db - job config state

### PR DESCRIPTION
**Pull request by gabb1er**
tbd.

### CODE REVIEW

File: checkita-core/src/main/resources/db/specific/h2/V1.1__create_job_state.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE "${defaultSchema}"."results_job_config"
+(
+    "job_id"            TEXT             NOT NULL,
```
Comment (gabb1er) [inline checkita-core/src/main/resources/db/specific/h2/V1.1__create_job_state.sql]: in H2 db change type for `job_id` to `VARCHAR(512)`

File: checkita-core/src/main/resources/db/specific/mssql/V1.1__create_job_state.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE "${defaultSchema}"."results_job_config"
+(
+    "job_id"            TEXT             NOT NULL,
+    "config"            TEXT             NOT NULL,
```
Comment (gabb1er) [inline checkita-core/src/main/resources/db/specific/mssql/V1.1__create_job_state.sql]: In MSSQL db type for `job_id` has to be `VARCHAR(512)` and the one for `config` column has to be `VARCHAR(MAX)`. 

Also need to change `TIMESTAMP` to `DATETIME`

File: checkita-core/src/main/resources/db/specific/mysql/V1.1__create_job_state.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE "${defaultSchema}"."results_job_config"
+(
+    "job_id"            TEXT             NOT NULL,
```
Comment (gabb1er) [inline checkita-core/src/main/resources/db/specific/mysql/V1.1__create_job_state.sql]: In MySQL db change type for `job_id` to `VARCHAR(512)`.

File: checkita-core/src/main/resources/db/specific/oracle/V1.1__create_job_state.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE "${defaultSchema}"."results_job_config"
+(
+    "job_id"            TEXT             NOT NULL,
```
Comment (gabb1er) [inline checkita-core/src/main/resources/db/specific/oracle/V1.1__create_job_state.sql]: In Oracle db change type for `job_id` to `VARCHAR(512)` and for `config` to `CLOB`

File: checkita-core/src/main/resources/db/specific/h2/V1.1__create_job_state.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE "${defaultSchema}"."results_job_config"
```
Comment (gabb1er) [inline checkita-core/src/main/resources/db/specific/h2/V1.1__create_job_state.sql]: Lets name this table as `job_state` (here and in files for other DBs) as it is not kind of result. I suppose that would be more appropriate as this table  will store job configuration states for which the results were computed.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 6]
```diff
@@ -49,6 +50,16 @@ object Models {
     val uniqueFieldNames: Seq[String] = Seq("jobId", "checkId", "referenceDate").map(camelToSnakeCase)
   }
 
+  sealed abstract class JobResult extends DQEntity {
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: Looks like this abstract class is overabundant: there is only one final case class that extends this one and we can extend it directly from DQEntity.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 7]
```diff
@@ -49,6 +50,16 @@ object Models {
...
+    val jobId: String
+    val config: String
+    val referenceDate: Timestamp
+    val executionDate: Timestamp
+
+    val uniqueFields: JobUniqueResult = jobId :: referenceDate :: HNil
+    val uniqueFieldNames: Seq[String] = Seq("jobId", "checkId", "referenceDate").map(camelToSnakeCase)
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: Remove `checkId`

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 8]
```diff
@@ -104,6 +115,15 @@ object Models {
                                    executionDate: Timestamp) extends CheckResult {
     val entityType: String = "loadCheckResult"
   }
+
+  final case class ResultJobConfig(
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: Inherit directly from `DQEntity` and rename to `JobState` in the same way as we would rename table.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 9]
```diff
@@ -132,6 +152,7 @@ object Models {
                                          numMetrics: Int,
                                          numChecks: Int,
                                          numLoadChecks: Int,
+                                         numJobConfig: Int,
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: I don't think we would ever need this metric. We always have sole instance of job configuration.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 10]
```diff
@@ -158,6 +179,7 @@ object Models {
                               checks: Seq[ResultCheck],
                               loadChecks: Seq[ResultCheckLoad],
                               metricErrors: Seq[ResultMetricErrors],
+                              jobConfig: Seq[ResultJobConfig],
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: Should be just `JobState`, not a sequence.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 11]
```diff
@@ -179,6 +201,7 @@ object Models {
               composedMetrics: Seq[ResultMetricComposed],
               checks: Seq[ResultCheck],
               loadChecks: Seq[ResultCheckLoad],
+              jobConfig: Seq[ResultJobConfig],
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: same as comment above.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala [hunk 12]
```diff
@@ -194,14 +217,15 @@ object Models {
         regularMetrics.size + composedMetrics.size,
         checks.size,
         loadChecks.size,
+        jobConfig.size,
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Models.scala]: remove

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala [hunk 13]
```diff
@@ -52,6 +53,17 @@ class Tables(val profile: JdbcProfile) {
     def getUniqueCond(r: R): Rep[Boolean] =
       jobId === r.jobId && checkId === r.checkId && referenceDate === r.referenceDate
   }
+
+  sealed abstract class JobResultTable[R <: ResultJobConfig](tag: Tag, schema: Option[String], table: String)
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala:57]: `JobStateTable[R <: JobState]`

Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala]: Looks like this abstract class is overabundant: there is only one class that extends this one and we can extend it directly from DQTable.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala [hunk 14]
```diff
@@ -138,6 +150,17 @@ class Tables(val profile: JdbcProfile) {
       executionDate
     ) <> (ResultCheckLoad.tupled, ResultCheckLoad.unapply)
   }
+
+  class ResultJobConfigTable(tag: Tag, schema: Option[String])
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala]: Rename to JobStateTable and inherit directly from DQTable. Change table name to `job_state` as per comments for db migration scripts.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala [hunk 15]
```diff
@@ -163,5 +186,12 @@ class Tables(val profile: JdbcProfile) {
       def getTableQuery(schema: Option[String]): TableQuery[ResultCheckLoadTable] =
         TableQuery[ResultCheckLoadTable]((t: Tag) => new ResultCheckLoadTable(t, schema))
     }
+
+    implicit object ResultJobConfigTableOps extends DQTableOps[ResultJobConfig] {
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/storage/Tables.scala]: Rename in the same way as other classes.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala [hunk 16]
```diff
@@ -293,6 +298,23 @@ trait DQJob extends Logging {
       metResults.toSeq.flatMap(_._2).flatMap(r => r.finalizeMetricErrors)
     }
 
+  protected def performJobResult(jobConfig: JobConfig)
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala]: Lets rename this method to `finalizeJobState`. In addition, do not wrap JobState in sequence.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala [hunk 17]
```diff
@@ -308,16 +330,17 @@ trait DQJob extends Logging {
   protected def combineResults(stage: String,
                                loadCheckResults: Seq[ResultCheckLoad],
                                checkResults: Result[Seq[ResultCheck]],
+                               jobResult: Result[Seq[ResultJobConfig]],
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala]: not a sequence

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala [hunk 18]
```diff
@@ -349,7 +372,8 @@ trait DQJob extends Logging {
             saveWithLogs(results.regularMetrics, "regular metrics"),
             saveWithLogs(results.composedMetrics, "composed metrics"),
             saveWithLogs(results.loadChecks, "load checks"),
-            saveWithLogs(results.checks, "checks")
+            saveWithLogs(results.checks, "checks"),
+            saveWithLogs(results.jobConfig, "job config")
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala]: change `job config` to `job state`

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala [hunk 19]
```diff
@@ -423,9 +447,10 @@ trait DQJob extends Logging {
     val regularMetricResults = finalizeRegularMetrics(getStage(storageStage), allMetricCalcResults)
     val composedMetricResults = finalizeComposedMetrics(getStage(storageStage), allMetricCalcResults)
     val metricErrors = finalizeMetricErrors(getStage(storageStage), allMetricCalcResults)
+    val jobResult = performJobResult(jobConfig)
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala]: change method name as per comment above

**Review by gabb1er — COMMENTED** [review]
Please, see my proposals in the comments.

File: checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala [hunk 20]
```diff
@@ -293,6 +297,38 @@ trait DQJob extends Logging {
...
+   * ready for writing into storage DB or sending via targets.
+   *
+   * @param jobConfig   Parsed Data Quality job configuration
+   * @param settings    Implicit application settings object
+   * @param jobId       Current Job ID
+   * @return Either a finalized of job state.
+   */
+  protected def finalizeJobState(jobConfig: JobConfig)
+                                (implicit settings: AppSettings, jobId: String): Result[JobState] = {
+
```
Comment (gabb1er) [inline checkita-core/src/main/scala/ru/raiffeisen/checkita/context/DQJob.scala:312]: Lets make it more Scala-stylish: 

```scala
val renderOpts = ...

val writeFunc = (jc: JobConfig) => settings.configEncryptor match {
    case Some(e) =>  writeEncryptedJobConfig(jc)(e)
    case None => writeJobConfig(jc)
}

wrtieFunc(jobConfig).map(jc => JobState(
  jobId,
  jc.root().render(renderOpts), 
  settings.referenceDateTime.getUtcTS,
  settings.executionDateTime.getUtcTS
))
```

File: docs/en/01-application-setup/01-ApplicationSettings.md [hunk 21]
```diff
@@ -75,7 +75,7 @@ Parameters for connecting to Data Quality results storage are defined in `storag
 For more information on results storage refer to [Data Quality Results Storage](03-ResultsStorage.md) chapter
 of the documentation.
 
-Thus, connection to storage is configured using following parameters:
```
Comment (gabb1er) [inline docs/en/01-application-setup/01-ApplicationSettings.md]: I believe, `Thus` was correct here.

File: docs/en/01-application-setup/01-ApplicationSettings.md [hunk 22]
```diff
@@ -122,6 +122,19 @@ It is also possible to provide list of default Spark configuration parameters us
...
+
+When `storage` section is defined, it is also recommended to use `encryption` section in order to protect sensitive 
+information in job config. This should be done by defining the parameters within the application configuration file:
+
+* `secret` -  Secret string used to encrypt/decrypt sensitive fields. This string should contain at least 32 characters. 
+*Required*. 
+* `keyFields` - List of key fields used to identify fields that requires encryption/decryption. 
+*Optional, default is `[password, secret]`*.
+
+If `encryption` section is missing then job config will be saved in database as is
```
Comment (gabb1er) [inline docs/en/01-application-setup/01-ApplicationSettings.md]: missing `.` at the end of sentece.

**Review by gabb1er — COMMENTED** [review]
Good job! :) Please, see my comment below.

**Review by cibaa-team-user — APPROVED** [review]

**Pull request merged**